### DISCUSSION TITLE
Exclude files that don't have docstrings

### DISCUSSION THREAD

**Discussion by KevinMusgrave** [mkdocstrings/mkdocstrings#412] [Ideas]
I'm using mkdocstrings with mkdocs-gen-files. The example in [the documentation](https://mkdocstrings.github.io/recipes) generates a page for every source file, even if there are no docstrings in that file. The files without docstrings get rendered as empty pages, so it'd be nice if either mkdocstrings or mkdocs-gen-files had an option to exclude these files.

**Comment by pawamoy** [comment]
Since you have control over the Python script that generates the pages using mkdocs-gen-files, I suggest that you simply prevent generation of a page based on an exclusion list.

Something like this:

```python
exclude_modules = {"src/project/lorem.py", "src/project/ipsum.py"}

for path in sorted(Path("src").rglob("*.py")):
    if path in exclude_modules:
        continue
    module_path = path.relative_to("src").with_suffix("")
    doc_path = path.relative_to("src").with_suffix(".md")
    full_doc_path = Path("reference", doc_path)
```

If you don't know in advance which module should be excluded, you can maybe use Griffe or directly visit the module's AST yourself to check if interesting members have docstrings or not.

If we ever create and provide a plugin that implement this recipe, we might consider adding this kind of options. In the meantime, it's out of scope (especially for mkdocs-gen-files). I'll also move this to discussions :slightly_smiling_face:

**Reply by KevinMusgrave** [reply]
Is there an easy way to determine if a .py file has any docstrings?

**Reply by pawamoy** [reply]
With Griffe :smile: 

```python
from griffe.loader import GriffeLoader
from griffe.collections import ModulesCollection
collection = ModulesCollection()
loader = GriffeLoader(modules_collection=collection)
loader.load_module("your_package")

for path in sorted(Path("src").rglob("*.py")):
    module_path = path.relative_to("src").with_suffix("")
    if not collection[module_path.parts].has_docstrings:
        continue
```

It means Griffe would be run twice for each build, though it's fast enough to be unnoticeable :slightly_smiling_face: 

Note: `.has_docstrings` checks the object and every member recursively. `.has_docstring` checks only the object itself.

**Reply by KevinMusgrave** [reply]
That works, thanks!

**Reply by oprypin** [reply]
> It means Griffe would be run twice for each build,

@pawamoy I wonder if this answer could be further improved by just accessing the actual parse tree that the collector already has.

In mkdocstrings-crystal this is an officially supported use case. E.g. https://github.com/athena-framework/website/blob/44c257cfc957c9d8864bac17326e33f6d663df3b/gen_doc_stubs.py#L6

**Comment by kroglkt** [comment]
Hello, I know this thread is old, but I have run into a very similar issue and would like to hear, if any changes have been made to mkdocstrings since then :-)

Is there a way to automatically see, if the MD file contains no text and if so, not display it? I cannot use a manual exclusion list for my case, because I am working on a template for other developers.

Thanks a lot!

**Reply by pawamoy** [reply]
As mentioned above, you can use `Object.has_docstrings` to check if the module you want to render on the page has any docstring (on itself or its members), to decide whether to create a page or not.
